File: android/core/src/main/kotlin/com/paladex/ex/core/net/Cache.kt
package com.paladex.ex.core.net

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * In-memory TTL cache for provider responses.
 *
 * Scraped pages are slow and rate-limited, so re-fetching one because the user
 * rotated the phone is not acceptable. Memory-only by design: on Android the
 * process outlives every screen that would want the data, and a price cache is
 * not worth persisting past a restart.
 */
class TtlCache(private val defaultTtlSeconds: Long = 3_600) {

    private data class Entry(val value: Any?, val expiresAt: Long)

    private val entries = mutableMapOf<String, Entry>()
    private val guard = Mutex()

    data class Result<T>(val value: T, val cached: Boolean)

    /** Run [block] unless a fresh cached value exists under [key]. */
    @Suppress("UNCHECKED_CAST")
    suspend fun <T> get(
        key: String,
        ttlSeconds: Long = defaultTtlSeconds,
        block: suspend () -> T,
    ): Result<T> {
        // A non-positive TTL means caching off, which is what tests want and
        // what makes scrape debugging bearable.
        if (ttlSeconds <= 0) return Result(block(), false)

        guard.withLock {
            entries[key]?.let { if (it.expiresAt > System.currentTimeMillis()) return Result(it.value as T, true) }
        }

        // Deliberately outside the lock: a slow scrape must not block reads of
        // unrelated keys. Two racing misses just fetch twice, which is fine.
        val value = block()

        guard.withLock {
            entries[key] = Entry(value, System.currentTimeMillis() + ttlSeconds * 1_000)
        }
        return Result(value, false)
    }

    suspend fun clear() = guard.withLock { entries.clear() }
}

File: src/lib/cache.ts
import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

import { env } from "./env";

/**
 * Two-tier cache: an in-process map for the hot path, backed by JSON files so
 * the cache survives dev-server reloads. Scraped pages are expensive and
 * rate-limited, so re-fetching them on every hot reload is not an option.
 */

interface Entry<T> {
  value: T;
  expiresAt: number;
}

const memory = new Map<string, Entry<unknown>>();
const CACHE_DIR = join(process.cwd(), ".cache");

function keyToFile(key: string): string {
  const hash = createHash("sha256").update(key).digest("hex").slice(0, 32);
  return join(CACHE_DIR, `${hash}.json`);
}

async function readDisk<T>(key: string): Promise<Entry<T> | null> {
  try {
    const raw = await readFile(keyToFile(key), "utf8");
    return JSON.parse(raw) as Entry<T>;
  } catch {
    return null;
  }
}

async function writeDisk<T>(key: string, entry: Entry<T>): Promise<void> {
  try {
    await mkdir(CACHE_DIR, { recursive: true });
    await writeFile(keyToFile(key), JSON.stringify(entry), "utf8");
  } catch {
    // A cache that cannot persist is still a working cache.
  }
}

/**
 * Run `fn` unless a fresh cached value exists. Returns the value plus whether
 * it came from cache, because the UI shows that distinction.
 */
export async function cached<T>(
  key: string,
  fn: () => Promise<T>,
  ttlSeconds = env.cacheTtlSeconds,
): Promise<{ value: T; cached: boolean }> {
  // A non-positive TTL means caching off, which is what tests want and what
  // makes SCRAPE-heavy debugging bearable.
  if (ttlSeconds <= 0) {
    return { value: await fn(), cached: false };
  }

  const now = Date.now();

  const hot = memory.get(key) as Entry<T> | undefined;
  if (hot && hot.expiresAt > now) {
    return { value: hot.value, cached: true };
  }

  const cold = await readDisk<T>(key);
  if (cold && cold.expiresAt > now) {
    memory.set(key, cold);
    return { value: cold.value, cached: true };
  }

  const value = await fn();
  const entry: Entry<T> = { value, expiresAt: now + ttlSeconds * 1000 };
  memory.set(key, entry);
  await writeDisk(key, entry);
  return { value, cached: false };
}

/** Drop everything. Exposed for the "refresh prices" button. */
export function clearMemoryCache(): void {
  memory.clear();
}
